src/alpaka/test/alpaka/HistoContainer_t.cc

```diff
@@ -0,0 +1,172 @@
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <limits>
+#include <random>
+
+#include "AlpakaCore/alpakaConfig.h"
+#include "AlpakaCore/alpakaWorkDivHelper.h"
+#include "AlpakaCore/HistoContainer.h"
+
+using namespace ALPAKA_ACCELERATOR_NAMESPACE;
+
+template <typename T>
+void go(const DevHost& host, const DevAcc1& device, Queue& queue) {
+  std::mt19937 eng;
+  std::uniform_int_distribution<T> rgen(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+
+  constexpr unsigned int N = 12000;
+  auto v_buf = alpaka::mem::buf::alloc<T, Idx>(host, N);
+  auto v = alpaka::mem::view::getPtrNative(v_buf);
+  auto v_d = alpaka::mem::buf::alloc<T, Idx>(device, N);
+  alpaka::mem::view::copy(queue, v_d, v_buf, N);
+
+  constexpr uint32_t nParts = 10;
+  constexpr uint32_t partSize = N / nParts;
+
+  using Hist = cms::alpakatools::HistoContainer<T, 128, N, 8 * sizeof(T), uint32_t, nParts>;
+  std::cout << "HistoContainer " << (int)(offsetof(Hist, off)) << ' ' << Hist::nbins() << ' ' << Hist::totbins() << ' '
+            << Hist::capacity() << ' ' << offsetof(Hist, bins) - offsetof(Hist, off) << ' '
+            << (std::numeric_limits<T>::max() - std::numeric_limits<T>::min()) / Hist::nbins() << std::endl;
+
+  auto offsets_buf = alpaka::mem::buf::alloc<uint32_t, Idx>(host, nParts + 1);
+  auto offsets = alpaka::mem::view::getPtrNative(offsets_buf);
+  auto off_d = alpaka::mem::buf::alloc<uint32_t, Idx>(device, nParts + 1);
+
+  auto h_buf = alpaka::mem::buf::alloc<Hist, Idx>(host, 1u);
+  auto h_d = alpaka::mem::buf::alloc<Hist, Idx>(device, 1u);
+
+  for (int it = 0; it < 5; ++it) {
+    offsets[0] = 0;
+    for (uint32_t j = 1; j < nParts + 1; ++j) {
+      offsets[j] = offsets[j - 1] + partSize - 3 * j;
+      assert(offsets[j] <= N);
+    }
+
+    if (it == 1) {  // special cases...
+      offsets[0] = 0;
+      offsets[1] = 0;
+      offsets[2] = 19;
+      offsets[3] = 32 + offsets[2];
+      offsets[4] = 123 + offsets[3];
+      offsets[5] = 256 + offsets[4];
+      offsets[6] = 311 + offsets[5];
+      offsets[7] = 2111 + offsets[6];
+      offsets[8] = 256 * 11 + offsets[7];
+      offsets[9] = 44 + offsets[8];
+      offsets[10] = 3297 + offsets[9];
+    }
+
+    alpaka::mem::view::copy(queue, off_d, offsets_buf, nParts + 1);
+
+    for (long long j = 0; j < N; j++)
+      v[j] = rgen(eng);
+
+    if (it == 2) {  // big bin
+      for (long long j = 1000; j < 2000; j++)
+        v[j] = sizeof(T) == 1 ? 22 : 3456;
+    }
+
+    alpaka::mem::view::copy(queue, v_d, v_buf, N);
+
+    alpaka::mem::view::set(queue, h_d, 0, 1u);
```

Comment:
Yes true, I have added it because when run several times in a loop, we want to be sure that the next iteration starts with sth clean.
This was also causing segfault.

Reply:
In practice, I would just declare `h_d` inside the loop (and in general obviously just keep variable declarations as close as possible to when they are used).
But here I would avoid changing the code in addition to the port to Alpaka, to be able to keep the portability 'portable'.